src/MagnumPlugins/UfbxImporter/UfbxImporter.cpp

```diff
@@ -0,0 +1,1162 @@
+                const ufbx_material_map &factorMap = list.maps[mapping.factorMap];
+                if (factorMap.has_value) {
+                    factor = Float(factorMap.value_real);
+                }
+            }
+
+            if (colorMap.has_value && mapping.attrib != MaterialAttribute{} && !addedAttributes[UnsignedInt(mapping.attrib)]) {
+                addedAttributes.set(UnsignedInt(mapping.attrib), true);
+                if (mapping.type == MaterialAttributeType::Float) {
+                    Float value = Float(colorMap.value_real) * factor;
+                    arrayAppend(attributes, {mapping.attrib, value});
+                } else if (mapping.type == MaterialAttributeType::Vector3) {
+                    Vector3 value = Vector3(colorMap.value_vec3) * factor;
+                    arrayAppend(attributes, {mapping.attrib, value});
+                } else if (mapping.type == MaterialAttributeType::Vector4) {
+                    Vector4 value = Vector4(colorMap.value_vec4) * Vector4{factor,factor,factor,1.0f};
+                    arrayAppend(attributes, {mapping.attrib, value});
+                } else {
+                    CORRADE_INTERNAL_ASSERT_UNREACHABLE(); /* LCOV_EXCL_LINE */
+                }
+            }
+
+            if (mapping.texture != MaterialAttribute{}) {
+                ufbx_texture *texture = colorMap.texture;
+                if (texture && colorMap.texture_enabled && !addedAttributes[UnsignedInt(mapping.texture)]) {
+                    addedAttributes.set(UnsignedInt(mapping.texture), true);
+                    arrayAppend(attributes, {mapping.texture, UnsignedInt(texture->typed_id)});
+
+                    if (mapping.textureMatrix != MaterialAttribute{} && texture->has_uv_transform) {
+
+                        /* Texture matrix should be connected to texture */
+                        CORRADE_INTERNAL_ASSERT(!addedAttributes[UnsignedInt(mapping.textureMatrix)]);
+                        addedAttributes.set(UnsignedInt(mapping.textureMatrix), true);
+
+                        const ufbx_matrix &mat = texture->uv_to_texture;
+                        Matrix3 value = {
+                            { Float(mat.m00), Float(mat.m10), 0.0f },
+                            { Float(mat.m01), Float(mat.m11), 0.0f },
+                            { Float(mat.m03), Float(mat.m13), 1.0f },
+                        };
+
+                        arrayAppend(attributes, {mapping.textureMatrix, value});
+                    }
+
+                    /* @todo Texture coordinates, either in ufbx or here */
+                }
+            }
+        }
+    }
+
+    if (material->features.double_sided.enabled) {
+        arrayAppend(attributes, {MaterialAttribute::DoubleSided, true});
+
+    }
+
+    /* Can't use growable deleters in a plugin, convert back to the default
+       deleter */
+    arrayShrink(attributes, DefaultInit);
+
+    MaterialTypes types;
+    types |= MaterialType::Phong;
```

Comment:
In theory the Phong values are always defined by property default values, but some materials omit them completely in the file (and don't even try to approximate whatever the fancier material model is) so maybe it's better to just have that if there's some actual data.

The question about unlit materials is a bit more tricky as FBX really only supports Phong, but everything else is written as custom properties that I add [here](https://github.com/bqqbarbhg/ufbx/blob/df8d5d6fccac23a5a0ae5382bd0047005fc6fb24/ufbx.c?ts=4#L12178-L12420) whenever I run into a new one. At least there is a standardized "Shader" system that is used *sometimes* :D But to the point: I have not seen anything write an unlit material, but I'd bet that 3ds Max would do it somehow as it tends to play very fast and loose with materials.

Reply:
Yep checked it quickly:

```
Material: 2622981919360, "Material::01 - Default", "" {
	Version: 102
	ShadingModel: "unknown"
	MultiLayer: 0
	Properties70:  {
		P: "ShadingModel", "KString", "", "", "unknown"
		P: "AmbientColor", "ColorRGB", "Color", "",0,0,0
		P: "DiffuseColor", "ColorRGB", "Color", "",0.75,0.75,0.75
		P: "SpecularColor", "ColorRGB", "Color", "",1,1,1
		P: "SpecularFactor", "double", "Number", "",1
		P: "ShininessExponent", "double", "Number", "",1024
		P: "TransparencyFactor", "double", "Number", "",0
		P: "EmissiveColor", "ColorRGB", "Color", "",0,0,0
		P: "EmissiveFactor", "double", "Number", "",0
		P: "3dsMax", "Compound", "", ""
		P: "3dsMax|ClassIDa", "int", "Integer", "",2121471519
		P: "3dsMax|ClassIDb", "int", "Integer", "",-1167678170
		P: "3dsMax|SuperClassID", "int", "Integer", "",3072
		P: "3dsMax|ai_matte Parameters/Connections", "Compound", "", ""
		P: "3dsMax|ai_matte Parameters/Connections|passthrough", "Reference", "", "A"
		P: "3dsMax|ai_matte Parameters/Connections|color", "ColorAndAlpha", "", "A",1,0,0,0
		P: "3dsMax|ai_matte Parameters/Connections|color.connected", "Bool", "", "A",1
		P: "3dsMax|ai_matte Parameters/Connections|color.shader", "Reference", "", "A"
		P: "3dsMax|ai_matte Parameters/Connections|opacity", "Color", "", "A",1,1,1
		P: "3dsMax|ai_matte Parameters/Connections|opacity.connected", "Bool", "", "A",1
		P: "3dsMax|ai_matte Parameters/Connections|opacity.shader", "Reference", "", "A"
		P: "3dsMax|ai_matte Results", "Compound", "", ""
	}
}
```

Not using the standard Shader system, embedding some sort of shader graph into the file, not even trying to approximate it with FBX properties, and even a completely opaque `ClassID`. So pretty much what you'd expect from FBX :D 

(I probably won't add this to the material library as 3ds Max has heaps of these and it would bloat the library unnecessarily to add every rare one)

Reply:
> FBX really only supports Phong

Great, that's the piece of info I was after :) For glTF it's the opposite, metallic/roughness is the only builtin material type (thus present in `MaterialType` always), others are optional extensions.

Could you add a comment there, about Phong being implicitly there because it's what FBX has builtin? 

> -1167678170

Wow, what a mess :scream: Given my [other comment](https://github.com/mosra/magnum-plugins/pull/133#discussion_r1032812952), is there any way for ufbx to just pass through the random unrecognized stuff? Probably out of scope for your library, I guess :D

Reply:
It's pretty trivial to pass those through, just need to avoid duplicating all the properly interpreted properties, but I think I have a solution for this (that ironically comes from solving material animations, which is a completely unrelated problem).